File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-8.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;

            if (nums[middle] >= nums[left]) {
                // This means nums[left..middle + 1] is sorted.
                if (nums[left] <= target && nums[middle] >= target) {
                    return searchIn(nums, left, middle, target);
                }

                left = middle + 1;
            } else {
                // This means nums[middle..right + 1] is sorted
                if (nums[middle] <= target && nums[right] >= target) {
                    return searchIn(nums, middle, right, target);
                }

                right = middle - 1;
            }
        }

        return -1;
    }

    private int searchIn(int[] nums, int leftBoundary, int rightBoundary, int target) {
        int left = leftBoundary;
        int right = rightBoundary;

        while (left <= right) {
            int middle = left + (right - left) / 2;

            if (nums[middle] < target) {
                left = middle + 1;
            } else if (nums[middle] > target) {
                right = middle - 1;
            } else {
                return middle;
            }
        }

        return -1;
    }
}

File: Data Structures & Algorithms/binary-search/submission-8.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def searchHelper(start, end):
            print(start, end)
            middle = (end + start) // 2
            print(middle)
            if end - start < 1 or nums[middle] == target:
                return middle if nums[middle] == target else -1
            if target > nums[middle]:
                return searchHelper(middle + 1, end)
            if target < nums[middle]:
                return searchHelper(start, middle - 1)
        return searchHelper(0, len(nums) - 1)